### PR TITLE
Flap Control related fixes

### PR DESCRIPTION
**Pull request by nikhar-abbas**
# Flap Control Related Fixes

## Purpose
Some updates to improve the optimization of the trailing edge flap controller gains. 

There were a number of issues to sort through as far as the DAC module was concerned. Primarily, error handling and management was improved for the xfoil calls. An added parameter is added to the modeling options so that we can directly pass the flapwise collective damping (structural+aero) into WEIS, as this is only available through an eigen-analysis (as far as I know). 

Additionally, MPI capabilities were improved for design of experiments. This is captured in the testing scripts, I think.

Just FYI, I ended up on a whirlwind of poor git management here. I think it's all cleaned up and makes sense, so hopefully I didn't cause too much confusion. 

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
The existing DAC tests should suffice

## Checklist
_Put an `x` in the boxes that apply._

- [x] I have run existing tests which pass locally with my changes
- [ ] I have added new tests or examples that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Comment by nikhar-abbas** [comment]
Unless I am missing something, it looks like `te_flaps_end` and `te_flaps_ext` were lost as design variables at some point, I added them back in in d683ae9

File: weis/glue_code/glue_code.py [hunk 1]
```diff
@@ -631,7 +631,7 @@ def setup(self):
             # Connections to turbine constraints
             if modeling_options['WISDEM']['TowerSE']['flag']:
                 self.connect('configuration.rotor_orientation', 'tcons_post.rotor_orientation')
-                self.connect('rlds_post.tip_pos.tip_deflection',     'tcons_post.tip_deflection')
+                self.connect('aeroelastic.max_TipDxc',          'tcons_post.tip_deflection')
```
Comment (ptrbortolotti) [inline weis/glue_code/glue_code.py:634]: I'm just confused by this, since earlier today I did merge the PR into weis develop...

Comment (nikhar-abbas) [reply weis/glue_code/glue_code.py]: Okay, this is fixed. I had cherry picked your commits before the PR was merged and that was causing confusing

**Comment by nikhar-abbas** [comment]
@johnjasa, @ptrbortolotti 

I think this is just about ready, but is still failing the tests. It looks like a ruamel.yaml error. I've seen this a bit, and it seems to show up when writing the yamls after a completed run. Have either of you seen this, or have an idea of how to solve it?

**Comment by johnjasa** [comment]
@nikhar-abbas, I haven't seen that ruamel error before. Do you have any leads for how to solve it, or are you still looking for solutions?
I know @ptrbortolotti has more experience with writing yamls than I do.

**Comment by ptrbortolotti** [comment]
we spot it and we have a dirty fix, Nikhar will try to find a cleaner way
it's all about writing out the modeling_options yaml...

**Comment by nikhar-abbas** [comment]
Okay, all of the examples are running locally now, but the tests are still failing. Looks like an endless wall of:
```
Warning: you have pip-installed dependencies in your environment file, but you do not list pip itself as one of your conda dependencies.  Conda may not use the correct pip to install your packages, and they may end up in the wrong place.  Please add an explicit pip dependency.  I'm adding one for you, but still nagging you.
```
@johnjasa is this a change you want me to make?

**Comment by gbarter** [comment]
I think you just have to add `pip` to the list of conda packages here:
https://github.com/nikhar-abbas/WEIS/blob/b/flp_fixes/environment.yml

Used to get the same error in WISDEM.  See the WISDEM `environment.yml` file here:
https://github.com/WISDEM/WISDEM/blob/master/environment.yml

**Comment by nikhar-abbas** [comment]
Okay, I think this is ready and the tests are passing! As far as running DOE's and optimizing flap specific parameters go, this branch seems to be working as expected.

One thing: I am getting a frame3DD related error when optimizing blade mass + flap controller gains. This error does not show up when optimizing blade mass alone, or flap related parameters alone. It looks like the forces are all becoming `NaNs` somehow (printing `nantest` [here](https://github.com/nikhar-abbas/WEIS/blob/e17f6cab60bb8565e600b79689beba5d1d2ed704/WISDEM/wisdem/pyframe3dd/pyframe3dd.py#L928) gives a large list of `True`), and I can't figure out why. If I add a print statement to print out the `exitCode`, I get the following error (cleaned up a bit for readability):
```
error in frame element property data: section property < 0 
Frame element number: 29  

error in x-axis trapezoidal loads, load case: 1 , element 1 , load 1
ending location = 3.442282 > L (0.000000) 

error: Non pos-def mass matrix
M[127][127] = 0.000000
error: Non pos-def mass matrix
M[128][128] = 0.000000
error: Non pos-def mass matrix
M[130][130] = 0.000000
error: Non pos-def mass matrix
M[131][131] = 0.000000
error: Non pos-def mass matrix
M[133][133] = 0.000000
error: Non pos-def mass matrix
M[134][134] = 0.000000
error: Non pos-def mass matrix
M[136][136] = 0.000000
error: Non pos-def mass matrix
M[137][137] = 0.000000

subspace: M[127][127] = 0.000000e+00 

exitCode = 229

Traceback (most recent call last):
  File "/home/nabbas/.conda-envs/weis-env/lib/python3.8/site-packages/openmdao/core/system.py", line 2171, in _call_user_function
    yield
  File "/home/nabbas/.conda-envs/weis-env/lib/python3.8/site-packages/openmdao/core/explicitcomponent.py", line 273, in _solve_nonlinear
    self.compute(self._inputs, self._outputs)
  File "/home/nabbas/WindEnergyToolbox/WEIS/WISDEM/wisdem/rotorse/rotor_structure.py", line 505, in compute
    displacements, forces, reactions, internalForces, mass, modal = blade.run()
  File "/home/nabbas/WindEnergyToolbox/WEIS/WISDEM/wisdem/pyframe3dd/pyframe3dd.py", line 931, in run
    raise RuntimeError("Frame3DD did not exit gracefully")
RuntimeError: Frame3DD did not exit gracefully
```

This does not show up until after a few iterations through WEIS, so I don't think it is MDAO-connection related or something like that. It may be related to my inputs, though they are consistent with other WEIS inputs that are working pretty well for me. Does anyone have any thoughts as to what might be causing this?

All things considered, this bug does seem to be pretty minor in the scheme of things, so I think this PR is ready for a more formal review.

**Comment by gbarter** [comment]
It looks like one of the blade elements in Frame3DD either has zero length or zero mass.  I'm not sure what is happening to the blade discretization when flaps are introduced.  There is a debug call in rotor_structure where we can write out the equivalent Frame3DD input file and can usually spot the issue.  I'm not sure what the optimizer is doing though to wind up in that state.

**Comment by ptrbortolotti** [comment]
can we merge @nikhar-abbas ? it's a green light from me

**Comment by nikhar-abbas** [comment]
@ptrbortolotti, I just pushed one more small fix. I don't expect it to cause the tests to fail, so once they all pass I think it is good to go

**Pull request merged**